pkg/ipcache/ipcache.go

```diff
@@ -0,0 +1,219 @@
+}
+
+// LookupByIP returns the corresponding security identity that endpoint IP maps
+// to within the provided IPCache, as well as if the corresponding entry exists
+// in the IPCache.
+func (ipc *IPCache) LookupByIP(endpointIP string) (identity.NumericIdentity, bool) {
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
+	identity, exists := ipc.ipToIdentityCache[endpointIP]
+	return identity, exists
+}
+
+// LookupByIdentity returns the set of endpoint IPs that have security identity
+// ID, as well as if the corresponding entry exists in the IPCache.
+func (ipc *IPCache) LookupByIdentity(id identity.NumericIdentity) (map[string]struct{}, bool) {
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
+	ips, exists := ipc.identityToIPCache[id]
+	return ips, exists
+}
+
+// IPIdentityMappingOwner is the interface the owner of an identity allocator
+// must implement
+type IPIdentityMappingOwner interface {
+	// OnIPIdentityCacheChange will be called whenever there the state of the
+	// IPCache has changed.
+	OnIPIdentityCacheChange()
+}
+
+// GetIPIdentityMapModel returns all known endpoint IP to security identity mappings
+// stored in the key-value store.
+func GetIPIdentityMapModel() {
+	// TODO (ianvernon) return model of ip to identity mapping. For use in CLI.
+	// see GH-2555
+}
+
+func ipIdentityWatcher(owner IPIdentityMappingOwner) {
+
+	for {
+		watcher := kvstore.ListAndWatch("endpointIPWatcher", IPIdentitiesPath, 512)
+
+		// Get events from channel as they come in.
+		for event := range watcher.Events {
+
+			var (
+				cacheChanged bool
+				ipIDPair     identity.IPIdentityPair
+			)
+
+			err := json.Unmarshal(event.Value, &ipIDPair)
+			if err != nil {
+				log.WithFields(logrus.Fields{"value": event.Value}).WithError(err).Errorf("not adding entry to ip cache; error unmarshaling data from key-value store")
+				continue
+			}
+
+			// Synchronize local caching of endpoint IP to ipIDPair mapping with
+			// operation key-value store has informed us about.
+
+			ipStr := ipIDPair.IP.String()
+
+			cachedIdentity, exists := IPIdentityCache.LookupByIP(ipStr)
+			endpointIPs, _ := IPIdentityCache.LookupByIdentity(ipIDPair.ID)
+
+			switch event.Typ {
+			case kvstore.EventTypeCreate, kvstore.EventTypeModify:
+				if !exists || cachedIdentity != ipIDPair.ID {
+					IPIdentityCache.upsert(ipStr, ipIDPair.ID)
+					cacheChanged = true
+
+					// Update XDS Cache as well.
+					ipStrings := make([]string, 0, len(endpointIPs))
+					for endpointIP := range endpointIPs {
+						ipStrings = append(ipStrings, endpointIP)
+					}
+					sort.Strings(ipStrings)
+					envoy.NetworkPolicyHostsCache.Upsert(envoy.NetworkPolicyHostsTypeURL, ipIDPair.ID.StringID(), &envoyAPI.NetworkPolicyHosts{Policy: uint64(ipIDPair.ID), HostAddresses: ipStrings}, false)
+				}
+			case kvstore.EventTypeDelete:
+				if exists {
+					// Delete from XDS Cache as well.
+					envoy.NetworkPolicyHostsCache.Delete(envoy.NetworkPolicyHostsTypeURL, cachedIdentity.StringID(), false)
```

Comment:
This is not correct. You only deleted one IP-ID pair. Not the whole identity.
You should delete form this cache only if there are no IPs left associated with an identity.
Otherwise, you must call **Upsert** with the set resulting from removing the IP in the deleted pair.